### PR TITLE
PXB-3059 add a innodb_redo_log archive option to xtrabackup

### PR DESCRIPTION
**Pull request by wangbincmss**
Backup utilities that copy redo log records may sometimes fail to keep pace with redo log generation while a backup operation is in progress, resulting in lost redo log records due to those records being overwritten.
If the MySQL server innodb_redo_log_archive_dirs =NULL and the redo log file storage media operates at a faster speed than the backup storage media,then result in backup fail.
Add a option --redo_log_arch_dir=tmp_redo_dir:redo_dir can solve this backup fail,
If innodb_redo_log_archive_dirs =NULL ,then after backup innodb_redo_log_archive_dirs is always NULL not change user setting.

### CODE REVIEW

**Comment by wangbincmss** [comment]
Add a option --redo_log_arch_dir

File: storage/innobase/xtrabackup/src/redo_log.cc [hunk 1]
```diff
@@ -953,6 +954,20 @@ void Archived_Redo_Log_Monitor::thread_func() {
     std::string dir;
     std::string subdir;
   } archive;
+  std::ostringstream query;
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/redo_log.cc]: move this declaration close to utilization

File: storage/innobase/xtrabackup/src/redo_log.cc [hunk 2]
```diff
@@ -953,6 +954,20 @@ void Archived_Redo_Log_Monitor::thread_func() {
...
+  read_mysql_variables(mysql, "SHOW VARIABLES", vars, true);
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/redo_log.cc]: is this necessary, having in mind we do the exact same query a few lines below ?

File: storage/innobase/xtrabackup/src/redo_log.h [hunk 3]
```diff
@@ -271,6 +271,9 @@ class Archived_Redo_Log_Monitor {
   /** readiness flag. */
   std::atomic<bool> ready;
 
+  /** redo arch flag. */
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/redo_log.h]: Elabore a bit more in the comment, also rename the variable name to something more useful. IIUC this flag controls if xtrabackup has set redo log arch

File: storage/innobase/xtrabackup/src/redo_log.cc [hunk 4]
```diff
@@ -953,6 +954,20 @@ void Archived_Redo_Log_Monitor::thread_func() {
...
+
+  if (!(redo_log_archive_dirs == nullptr || *redo_log_archive_dirs == 0)) {
+    xb::info() << "innodb_redo_log_archive_dirs is now  set .";
+    redo_arch = true;
+  }
+
+  if (xtrabackup_redo_log_arch_dir) {
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/redo_log.cc]: Here we are setting redo archive unconditionally, what happens if someone already have it set? we will overwrite to --redo-log-arch-dir and later return it back to NULL. We should only set if it is not already set.

**Review by altmannmarcelo — CHANGES_REQUESTED** [review]
Hi @wangbincmss .

Thanks a lot for your contribution. I have left a few comments to improve your PR. Apart from that I would like to ask you to enhance in two areas:
1) We try to avoid making xtrabackup change server configuration. When we do so, we need to ensure that in case of error we return the configuration back to the original value. Can you please add some error handling that if xtrabackup fails after we setting redo log archive we restore the configuration to its original value?

2) We have a test suite that ensures we are not introducing regression on other parts of the code, can you please create a test case to cover the funtionality we are adding here ? You can check https://github.com/percona/percona-xtrabackup/blob/8.0/storage/innobase/xtrabackup/test/t/xb_log_archiving.sh and https://github.com/percona/percona-xtrabackup/blob/8.0/storage/innobase/xtrabackup/test/inc/xb_log_archiving.sh

**Review by altmannmarcelo — CHANGES_REQUESTED** [review]
@wangbincmss I think you re-requested the review by mistake, please have a look on above comments.

File: storage/innobase/xtrabackup/src/redo_log.cc (hunk 2 above)
Comment (wangbincmss) [reply storage/innobase/xtrabackup/src/redo_log.cc]: Yes，need to check   SET GLOBAL innodb_redo_log_archive_dirs  takes effect or not

**Comment by wangbincmss** [comment]
@altmannmarcelo hi ,    I make some changes
1: add some error handling that if xtrabackup fails,  the server option changes only when  redo log archive is null , so error handling : changing redo log archive  to original value (null)  if xtrabackup fails
2:  avoid making xtrabackup change server configuration.: the server option changes only when  redo log archive is null, and bakup finished then change  to original value (null)
3:     rename the variable  to innodb_redo_log_archive   and   add a bit   comments
4: add four test cases    covers   server  and xtrabackup redo log  redo  option is NULL and true value

File: storage/innobase/xtrabackup/src/redo_log.cc [hunk 5]
```diff
@@ -953,17 +954,31 @@ void Archived_Redo_Log_Monitor::thread_func() {
...
     std::string subdir;
   } archive;
-
+  
   read_mysql_variables(mysql, "SHOW VARIABLES", vars, true);
 
-  if (redo_log_archive_dirs == nullptr || *redo_log_archive_dirs == 0) {
+  if (!(redo_log_archive_dirs == nullptr || *redo_log_archive_dirs == 0)) {
+    xb::info() << "innodb_redo_log_archive_dirs is already  set .";
+    innodb_redo_log_archive = true;
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/redo_log.cc]: what the code is doing and what the description of the variable says is contradicting.
You set this to true if the redo archive is already set on the server. Based on your comment this should be true when PXB changed redo archive dir from null to xtrabackup_redo_log_arch_dir.

File: storage/innobase/xtrabackup/src/redo_log.cc [hunk 6]
```diff
@@ -1014,6 +1029,10 @@ void Archived_Redo_Log_Monitor::thread_func() {
 
   if (res == nullptr) {
     xb::info() << "Redo Log Archiving is not used.";
+    if (!innodb_redo_log_archive) {
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/redo_log.cc]: Since we are repeating the same code in multiple places, we might create a member function to execute the cleanup.

File: storage/innobase/xtrabackup/src/redo_log.cc [hunk 7]
```diff
@@ -1128,6 +1161,11 @@ void Archived_Redo_Log_Monitor::thread_func() {
   }
 
   xb_mysql_query(mysql, "SELECT innodb_redo_log_archive_stop()", false);
+  if (!innodb_redo_log_archive) {
+    xb::info() << "innodb_redo_log_archive_dirs is not set,set it NULL .";
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/redo_log.cc]: This message is measleading, if it is not set already (I read this as NULL) why we set it to NULL again ?

File: storage/innobase/xtrabackup/src/xtrabackup.cc [hunk 8]
```diff
@@ -849,6 +851,11 @@ struct my_option xb_client_options[] = {
      (G_PTR *)&xtrabackup_incremental_basedir,
      (G_PTR *)&xtrabackup_incremental_basedir, 0, GET_STR, REQUIRED_ARG, 0, 0,
      0, 0, 0, 0},
+    {"redo_log_arch_dir", OPT_INNODB_REDO_LOG_ARCHIVE_DIRS,
+     "redo log archive destination directory",
+     (G_PTR *)&xtrabackup_redo_log_arch_dir,
+     (G_PTR *)&xtrabackup_redo_log_arch_dir, 0, GET_STR, OPT_ARG, 0, 0, 0, 0, 0,
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/xtrabackup.cc]:
```suggestion
     (G_PTR *)&xtrabackup_redo_log_arch_dir, 0, GET_STR, REQUIRED_ARG, 0, 0, 0, 0, 0,
```

Comment (altmannmarcelo) [reply storage/innobase/xtrabackup/src/xtrabackup.cc]: This is not a flag, we need to require argument

File: storage/innobase/xtrabackup/src/redo_log.cc (hunk 5 above)
Comment (altmannmarcelo) [reply storage/innobase/xtrabackup/src/redo_log.cc]: I would rather re-work this whole block. Its a bit counter intuitive. Some thing like below will be better:

```
if (redo_log_archive_dirs == nullptr || *redo_log_archive_dirs == 0) {
  std::ostringstream query;
  if (xtrabackup_redo_log_arch_dir) {
    query << "SET GLOBAL innodb_redo_log_archive_dirs ="
          << "\"" << xtrabackup_redo_log_arch_dir << "\"";
    xb_mysql_query(mysql, query.str().c_str(), false, true); <-- we need to keep this to fail in case of error
    innodb_redo_log_archive = true;
  } else {
    xb::info() << "Redo Log Archiving is not set up.";
    free_mysql_variables(vars);
    mysql_close(mysql);
    my_thread_end();
    return;
  }
}
xb::info() << "xtrabackup redo_log_arch_dir is now  set .";
```

**Review by altmannmarcelo — CHANGES_REQUESTED** [review]

File: storage/innobase/xtrabackup/src/redo_log.cc (hunk 7 above)
Comment (wangbincmss) [reply storage/innobase/xtrabackup/src/redo_log.cc]: feels a little measleading ，the true meaning is  the server innodb_redo_log_archive_dirs is not set(NULL) and PXB innodb_redo_log_archive_dirs is not NULL .  in order to not change user option, so change  it to NULL at last.
 I will  correct  the mistake.

File: storage/innobase/xtrabackup/src/redo_log.cc (hunk 6 above)
Comment (wangbincmss) [reply storage/innobase/xtrabackup/src/redo_log.cc]: I will create it
